### PR TITLE
A few dedicated exhaust related issues

### LINKED ISSUE #9347: AirLoopHVAC:ExhaustSystem fan schedule warning
Issue overview
--------------
Running 5ZoneAirCooledWithDOASAirLoop_ExhaustSystem.idf results in these warnings:
```
   ** Warning ** GetExhaustAirSystemInput: AirLoopHVAC:ExhaustSystem=EXHAUSTSYSTEM1
   **   ~~~   ** Could not find a match for Central Exhaust Fan's Avaiability Schedule.
   **   ~~~   ** It will be treated as always available.
   ** Warning ** GetExhaustAirSystemInput: AirLoopHVAC:ExhaustSystem=EXHAUSTSYSTEM2
   **   ~~~   ** Could not find a match for Central Exhaust Fan's Avaiability Schedule.
   **   ~~~   ** It will be treated as always available.
```

This warning isn't really necessary and could be deleted.

If there is an argument for keeping this, then 
* there is a typo "Avaiability" 
* the example file should be modified so that it does not produce this warning.

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 22.1.0-4ba5fb374c

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file 5ZoneAirCooledWithDOASAirLoop_ExhaustSystem
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jcyuan2020**
Pull request overview
---------------------
A few changes related to Dedicated Exhaust System:
 - Fixes #9347; also some changes in the ZoneHVAC:ExhaustControl's availability schedule handling as well.
 - Added a supply node check function to check the supply or supply list input field for ZoneHVAC:ExhaustControl object. This will make the input processing safer by checking the input field is really a zone inlet---originally this relies on the user to assure the correctness of the input; 
 - A unit test is added to test the newly added supply node function;
 - Switch to enum class on of one choice variable;
 - Some additional minor refactoring/clean up, e.g. use class members' initialization at definition instead of a constructor.

Diffs: generally no diffs are expected, expect the text diff in 5ZoneAirCooledWithDOASAirLoop_ExhaustSystem.idf. The new version should be void of the following lines:
```
   ** Warning ** GetExhaustAirSystemInput: AirLoopHVAC:ExhaustSystem=EXHAUSTSYSTEM1
   **   ~~~   ** Could not find a match for Central Exhaust Fan's Avaiability Schedule.
   **   ~~~   ** It will be treated as always available.
   ** Warning ** GetExhaustAirSystemInput: AirLoopHVAC:ExhaustSystem=EXHAUSTSYSTEM2
   **   ~~~   ** Could not find a match for Central Exhaust Fan's Avaiability Schedule.
   **   ~~~   ** It will be treated as always available.
```
**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/ExhaustAirSystemManager.cc [hunk 1]
```diff
@@ -502,20 +492,19 @@ namespace ExhaustAirSystemManager {
                 ip->markObjectAsUsed(cCurrentModuleObject, instance.key());
 
                 std::string availSchName = ip->getAlphaFieldValue(objectFields, objectSchemaProps, "availability_schedule_name");
```
Comment (amirroth) [inline src/EnergyPlus/ExhaustAirSystemManager.cc:493]: 📌 How many times does the string value get copied in `getAlphaFieldValue`?

File: src/EnergyPlus/ExhaustAirSystemManager.cc [hunk 2]
```diff
@@ -580,6 +571,7 @@ namespace ExhaustAirSystemManager {
                 thisExhCtrl.ExhaustFlowFractionScheduleNum = exhaustFlowFractionScheduleNum;
 
                 std::string supplyNodeOrNodelistName = ip->getAlphaFieldValue(objectFields, objectSchemaProps, "supply_node_or_nodelist_name");
+                thisExhCtrl.SupplyNodeOrNodelistName = supplyNodeOrNodelistName;
```
Comment (amirroth) [inline src/EnergyPlus/ExhaustAirSystemManager.cc]: You may be doing an extra deep copy by creating a temporary variable here.

File: src/EnergyPlus/ExhaustAirSystemManager.cc [hunk 3]
```diff
@@ -712,12 +704,14 @@ namespace ExhaustAirSystemManager {
         // Calculate a zonehvac exhaust control system
 
         auto &thisExhCtrl = state.dataZoneEquip->ZoneExhaustControlSystem(ZoneHVACExhaustControlNum);
+
         int InletNode = thisExhCtrl.InletNodeNum;
         int OutletNode = thisExhCtrl.OutletNodeNum;
         auto &thisExhInlet = state.dataLoopNodes->Node(InletNode);
         auto &thisExhOutlet = state.dataLoopNodes->Node(OutletNode);
         Real64 MassFlow = state.dataLoopNodes->Node(InletNode).MassFlowRate;
```
Comment (amirroth) [inline src/EnergyPlus/ExhaustAirSystemManager.cc]: You've already defined `thisExhInlet`, use it.

File: src/EnergyPlus/ExhaustAirSystemManager.cc [hunk 4]
```diff
@@ -898,10 +890,40 @@ namespace ExhaustAirSystemManager {
...
+    void CheckForSupplyNode(EnergyPlusData &state, int const ExhCtrlNum, bool &NodeNotFound)
     {
         // Trying to check a node to see if it is truely a supply node
         // for a nodelist, need a call loop to check each node in the list
+
+        auto &thisExhCtrl = state.dataZoneEquip->ZoneExhaustControlSystem(ExhCtrlNum);
+
+        // check a node is a zone inlet node.
+        std::string_view RoutineName = "GetExhaustControlInput: ";
+        std::string CurrentModuleObject = "ZoneHVAC:ExhaustControl";
```
Comment (amirroth) [inline src/EnergyPlus/ExhaustAirSystemManager.cc]: Can this not also be a `string_view`?

File: src/EnergyPlus/ExhaustAirSystemManager.hh [hunk 5]
```diff
@@ -62,68 +62,61 @@ namespace ExhaustAirSystemManager {
...
-        Real64 DesignExhaustFlowRate;
-        int FlowControlTypeNum;
-        int ExhaustFlowFractionScheduleNum;
-        int SupplyNodeOrNodelistNum;
-        int MinZoneTempLimitScheduleNum;
-        int MinExhFlowFracScheduleNum;
-        int BalancedExhFracScheduleNum;
-        Real64 BalancedFlow;
-        Real64 UnbalancedFlow;
+        enum class FlowControlType
```
Comment (amirroth) [inline src/EnergyPlus/ExhaustAirSystemManager.hh:89]: What is our convention for declaring these enumerations inside the `struct`/`class` vs. just inside the namespace?

File: tst/EnergyPlus/unit/ExhaustSystem.unit.cc [hunk 6]
```diff
@@ -424,3 +429,151 @@ TEST_F(EnergyPlusFixture, ExhaustSystemInputTest)
...
+    state->dataGlobal->NumOfZones = 4;
+    state->dataHeatBal->Zone.allocate(state->dataGlobal->NumOfZones);
+    state->dataHeatBal->Zone(1).Name = "ZONE1";
+    state->dataHeatBal->Zone(2).Name = "ZONE2";
+    state->dataHeatBal->Zone(3).Name = "ZONE3";
+    state->dataHeatBal->Zone(4).Name = "ZONE4";
+
+    state->dataSize->FinalZoneSizing.allocate(4);
+    state->dataSize->FinalZoneSizing(2).MinOA = 0.25;
+
```
Comment (amirroth) [inline tst/EnergyPlus/unit/ExhaustSystem.unit.cc:445]: `auto &thisX` is good in unit tests also.

File: src/EnergyPlus/ExhaustAirSystemManager.cc (hunk 1 above)
Comment (jcyuan2020) [reply src/EnergyPlus/ExhaustAirSystemManager.cc:493]: It seems at least the last parameter of this function can be passed as `string_view`:
```
std::string InputProcessor::getAlphaFieldValue(json const &ep_object, json const &schema_obj_props, std::string const &fieldName)
{
    // Return the value of fieldName in ep_object as a string.
    // If the field is not present in ep_object then return its default if there is one, or return an empty string
    auto const &schema_field_obj = schema_obj_props[fieldName];
```

The local string `value` needs to be a writable one, so probably nothing could do on it: 
```
    std::string value;
    auto it = ep_object.find(fieldName);
    if (it != ep_object.end()) {
        auto const &field_value = it.value();
        if (field_value.is_string()) {
            auto valuePair = getObjectItemValue(field_value.get<std::string>(), schema_field_obj);
            value = valuePair.first;
            isDefaulted = valuePair.second;
        } else {
            assert(false); // String value requested but field type is numeric
        }
    } else {
        isDefaulted = findDefault(value, schema_field_obj);
        if (!isDefaulted) {
            value = "";
```

File: src/EnergyPlus/ExhaustAirSystemManager.cc (hunk 2 above)
Comment (jcyuan2020) [reply src/EnergyPlus/ExhaustAirSystemManager.cc]: Thanks! Removed the temp variable and now it uses the struct member directly.

File: src/EnergyPlus/ExhaustAirSystemManager.cc (hunk 3 above)
Comment (jcyuan2020) [reply src/EnergyPlus/ExhaustAirSystemManager.cc]: Thanks! Now it is:
```
        Real64 MassFlow = thisExhInlet.MassFlowRate;
```

File: src/EnergyPlus/ExhaustAirSystemManager.hh (hunk 5 above)
Comment (jcyuan2020) [reply src/EnergyPlus/ExhaustAirSystemManager.hh:89]: Probably it does not have to be in one particular way over the other. But in general, I found it seems a reasonable balance to have the enum definition placed inside the struct that it is to be used, usually in the `.hh` file; and then have the string_view _UCs placed just inside the name space, usually in the `.cc` file to limit scope. The former will help to keep it in necessary scope and easy to find or being moved around with the entire struct in one piece. The latter will help to limit the reduce unnecessary visibility to this particular `.cc` file.

File: src/EnergyPlus/ExhaustAirSystemManager.cc (hunk 4 above)
Comment (jcyuan2020) [reply src/EnergyPlus/ExhaustAirSystemManager.cc]: Changed to `string_view`.

File: tst/EnergyPlus/unit/ExhaustSystem.unit.cc (hunk 6 above)
Comment (jcyuan2020) [reply tst/EnergyPlus/unit/ExhaustSystem.unit.cc:445]: Thanks! Use the shortcut on the new unit test.

**Comment by Myoldmopar** [comment]
This is showing no diffs.  In fact the error file diffs that you mentioned in the top post aren't showing up as diffs.  That's ... good?

**Comment by jcyuan2020** [comment]
It cannot be good if there is no diff on that example file (5ZoneAirCooledWithDOASAirLoop_ExhaustSytem.idf). However, I think I found the diffs were reported as expected:
https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-dbc74f291d9f83d9da69a168024e4e5b532ecee9-PR9665-x86_64-MacOS-10.17-clang-13.0.0.html

and here:
https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-dbc74f291d9f83d9da69a168024e4e5b532ecee9-PR9665-x86_64-Linux-Ubuntu-18.04-gcc-7.5.html 

So it should still be good now.

**Comment by Myoldmopar** [comment]
Oh good catch!! I usually rely on the dashboard showing 100% vs 99%.  "Unfortunately", we are at a point in the test count where one regression can go by and it rounds up to 100%.  I'll have to be a little more careful.  It would be great if we could re-enable GitHub showing commit comments on the PR like it used to, because then we'd see the full pass/fail count right here.  Anyway, I assume this is ready to go in then and I'll work to get it merged.

File: src/EnergyPlus/ExhaustAirSystemManager.cc [hunk 7]
```diff
@@ -599,15 +588,16 @@ namespace ExhaustAirSystemManager {
                                               DataLoopNode::ConnectionType::Sensor,
                                               NodeInputManager::CompFluidStream::Primary,
                                               DataLoopNode::ObjectIsNotParent); // , // _, // supplyNodeOrNodelistName);
-                thisExhCtrl.SupplyNodeOrNodelistNum = supplyNodeOrNodelistNum;
+
                 // Verify these nodes are indeed supply nodes:
                 bool nodeNotFound = false;
-                if (thisExhCtrl.FlowControlTypeNum == 1) { // FollowSupply
+                if (thisExhCtrl.FlowControlOption == ZoneExhaustControl::FlowControlType::FollowSupply) { // FollowSupply
```
Comment (Myoldmopar) [inline src/EnergyPlus/ExhaustAirSystemManager.cc:594]: Just so much better...good stuff here!

File: src/EnergyPlus/ExhaustAirSystemManager.hh (hunk 5 above)
Comment (Myoldmopar) [reply src/EnergyPlus/ExhaustAirSystemManager.hh:89]: :+1:

File: src/EnergyPlus/ExhaustAirSystemManager.hh [hunk 8]
```diff
@@ -62,68 +62,61 @@ namespace ExhaustAirSystemManager {
...
-        Real64 centralFan_MassFlowRate;
-        Real64 centralFan_VolumeFlowRate_Std;
-        Real64 centralFan_VolumeFlowRate_Cur;
-        Real64 centralFan_Power;
-        Real64 centralFan_Energy;
+        Real64 centralFan_MassFlowRate = 0.0;
+        Real64 centralFan_VolumeFlowRate_Std = 0.0;
+        Real64 centralFan_VolumeFlowRate_Cur = 0.0;
+        Real64 centralFan_Power = 0.0;
+        Real64 centralFan_Energy = 0.0;
```
Comment (Myoldmopar) [inline src/EnergyPlus/ExhaustAirSystemManager.hh:81]: Simple cleanup, but effective.

File: tst/EnergyPlus/unit/ExhaustSystem.unit.cc [hunk 9]
```diff
@@ -424,3 +429,156 @@ TEST_F(EnergyPlusFixture, ExhaustSystemInputTest)
...
+                                                                    DataLoopNode::ObjectIsParent);
+
+    NodeNotFound = false;
+    ExhaustAirSystemManager::CheckForSupplyNode(*state, ExhCtrlNum, NodeNotFound);
+
+    EXPECT_TRUE(NodeNotFound);
+
+    EXPECT_EQ(state->dataErrTracking->TotalWarningErrors, 0);
+    EXPECT_EQ(state->dataErrTracking->TotalSevereErrors, 1);
+    EXPECT_EQ(state->dataErrTracking->LastSevereError, "GetExhaustControlInput: ZoneHVAC:ExhaustControl=");
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/ExhaustSystem.unit.cc:583]: Nice and clean.  Good stuff.

**Review by Myoldmopar — APPROVED** [review]
LGTM!

**Comment by Myoldmopar** [comment]
Building and testing locally now.

**Comment by Myoldmopar** [comment]
Everything happy here, reproduced the same ERR diff in that single file.  This is ready to go, thanks @jcyuan2020

**Pull request merged**